README.md
# MPid

A microcontroller-friendly PID module.	

[![Build Status](https://travis-ci.org/gbmhunter/MPid.png?branch=master)](https://travis-ci.org/gbmhunter/MPid)

## Description

A light-weight, fast PID library designed for use on embedded systems (but can also run on any machine which has a G++ compiler).

For better performance and control, the PID library supports a generic number type. Number type must support casting to doubles. Fixed-point numbers are recommended for high-speed operation, doubles are recommended for non-time critical algorithms.

Relies on used calling `Pid.Run()` at a regular and fixed interval (usually in the milli-second range, via an interrupt).

Automatically adjusts `Kp`, `Ki` and `Kd` depending on the chosen time step (`Zp`, `Zi`, and `Zd` are the time-step adjusted values).

**DO NOT** try and make `Kp`, `Ki`, or `Kd` negative. This results in undefined behaviour.

### Smooth Control

Derivative control is only active when at least two calls to `Run()` have been made (does not assume previous input was 0 on first call, which can cause a huge derivative jolt!).

### Easy Debugging

You can print PID debug information by providing a callback via :code:`Pid::SetDebugPrintCallback()`, which supports method callbacks by utilizing the slotmachine-cpp library. 

The following code shows you how to assign a callback for debug printing.

```c++
class Printer {
	public:
		void PrintDebug(const char* msg) {
			std::cout << msg;
		}
};

Printer myPrinter;
Pid pidTest;

// Asign callback to Printer's PrintDebug function.
this->pidTest.SetDebugPrintCallback(SlotMachine::CallbackGen<Printer, void, const char*>(&myPrinter, &Printer::PrintDebug));
```

You can disable all debug info (to free up some memory space) by setting `cp3id_config_INCLUDE_DEBUG_CODE` in `include/Config.hpp` to `0`. The debug buffer size can be changed with `cp3id_config_DEBUG_BUFF_SIZE`, again in `Config.hpp`.

## Code Dependencies


| Dependency          |   Delivery            | Usage
|---------------------|-----------------------|--------------------------------------------
| `<cstdint>`         |  Standard C++ library | Fixed-width variable type definitions (e.g. `uint32_t`).
| MAssert             |  External module      | Providing runtime safety checks against this module.
| MUnitTest           |  External module      | Framework for unit tests.


## Building

This example assumes you are running on a Linux-like system and have basic build tools like `gcc` and `make` installed, as well as `cmake`.

```sh
$ git clone https://github.com/gbmhunter/MFixedPoint.git
$ cd MFixedPoint
$ mkdir build
$ cd build
$ cmake ..
$ make
```

## Usage

```c++
// Create a PID object which uses double for all of it's calculations, inputs, and outputs.
Pid<double> pidTest;

main() {
	// Set-up PID controller, non accumulating
	pidTest.Init(
		1.0,									//!< Kp
		1.0,									//!< Ki
		1.0,									//!< Kd
		Pid<double>::PID_DIRECT,				//!< Control type
		Pid<double>::DONT_ACCUMULATE_OUTPUT,	//!< Control type
		10.0,									//!< Update rate (ms)
		-100.0,									//!< Min output
		100.0,									//!< Max output
		0.0										//!< Initial set-point
	);
}

// Call every 10.0ms (as set in Pid.Init())
TimerIsr() {
	// Read input
	input = ReadPin(2);
	// Perform one execution
	pidTest.Run(input);
	// Set output
	SetPin(3) = pidTest.output;
}
```

See `test/PidTest.cpp` for more examples.
	
## Issues


For known bugs, desired enhancements e.t.c, see GitHub issues section.
	
## Changelog

See `CHANGELOG.md`.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html) (as of v7.0.0).

## [Unreleased]

- Nothing here yet.

## [v5.0.0] - 2019-05-20

### Added

- Added .gitingore file.
- Added `LICENSE.txt` (MIT license).

### Changed

- Replaced Make build script with CMake.
- Replace the custom four number versioning with the standarized semantic verioning (three number).
- Changed README from ReStructuredText to Markdown format.

## Old Style Changelog (pre v5.0.0)

========= ========== ===================================================================================================
Version   Date       Comment
========= ========== ===================================================================================================
v4.0.1.0  2014-10-11 Updated .travis.yml file to support C++11, closes #40.
v4.0.0.0  2014-10-10 Replaced UnitTest++ with the MUnitTest module, closes #32. Changed name to MPid, closes #31. Wrapped all code in MbeddedNinja namespace, closes #33. Removed slotmachine-cpp from 'lib/', closes #34. Updated Makefile to be similar to all of the other MToolkit modules, closes #35. Added main.cpp file for tests, closes #36. Moved all Pid::Init() code to constructor, closes #37.
v3.1.2.3  2014-04-16 Fixed up some code styling in the README. Added code example for assiging callback for debug printing, closes #30.
v3.1.2.2  2014-04-16 Properly styled code in README, closes #29. Added slotmachine-cpp to list of dependencies in README, closes #27.
v3.1.2.1  2014-04-16 Attempt to fix identation issue in 'Usage' section of README, closes #28.
v3.1.2.0  2014-03-24 Fixed the git submodule address for slotmachine-cpp so it was public, now TravisCI should be able to access it, closes #26. 
v3.1.1.0  2014-03-24 Changed typedef enums into enum classes and added compiler flag -std=c++0x to Makefile, closes #18. Updated unit test file accordingly.
v3.1.0.0  2014-03-24 Added include/Config.hpp to hold all PID configuration settings. Renamed debug print macros to fit. Added slotmachine-cpp lib to lib/slotmachine-cpp, added callback functionality for printing debug information, closes #25. Fixed incorrect checking of the size of the debug buffer, closes #24. Added relevant info to README.
v3.0.2.0  2014-03-24 Renamed actualKp and associated variables to just Kp, as Zp is now used to describe the time-adjusted ones, closes #2. Added comment to example code in README. Pid::Run() now calculates pTerm correctly, closes #3. Fixed improper use of carat symbol in Pid::Run() bug, closes #23.
v3.0.1.0  2014-03-24 Added this keyword to Init() function (and others) to prevent self-assignment bug, closes #19. Removed all references of the fixed-point library, now that the library accepts a dynamic data type, closes #20. Changed namespace from PidNs to CP3id, closes #22. Add more comments to PID functions. Updated README project title.
v3.0.0.0  2014-03-24 Moved src/include/ to include/. Added Eclipse C++ project files. Added api/CP3id.hpp, which is a single file you can include to use the CP3id library.
v2.5.1.0  2013-08-29 Cleaned up Makefile.
v2.5.0.1  2013-08-29 Fixed broken Travis CI build status image in README.
v2.5.0.0  2013-08-29 Added dependency awareness to Makefile, it now re-builds on header file changes.
v2.4.1.0  2013-08-28 Updated and fixed the build issue with the Makefile.
v2.4.0.3  2013-08-27 Changed reference to BitBucket in README to GitHub.
v2.4.0.2  2013-08-27 Changed README Changelog to table format.
v2.4.0.1  2013-08-27 Added Travis CI build status image that I had forgotten to do in the last commit.
v2.4.0.0  2013-08-27 Added .travis.yml file in root directory for Travis CI compatibility. Added Travis CI build status image to top of README.
v2.3.0.0  2013-05-31 Renamed README.md back to README.rst. Added Makefile in root dir that compiles everything and runs unit tests. Added note about Cpp-Pid superseding C-Pid. Removed object and .d files.
v2.2.2.1  2013-05-28 Added usage example to README.md.
v2.2.2.0  2013-05-28 Renamed namespace from Pid to PidNs and Pid class from PidDbl (no longer relevant with templating) to Pid.
v2.2.1.0  2013-05-28 Changed zKp, zKi, zKd (time-scaled constants) to Zp, Zi, Zd. Removed incorrect include directory in test/Makefile.
v2.2.0.0  2013-05-28 Removed specific fixed-point implementation in Pid.cpp, with the idea that you can now use the template code.
v2.1.0.0  2013-05-28 Converted more README sections to markdown format. Added Makefile for unit tests.
v2.0.0.3  2013-05-27 Fixed version numbers.
v2.0.0.2  2013-05-27 Changed README.rst to README.md (to markdown format).
v2.0.0.1  2013-05-27 Fixed commit issues, removed .goutstream files.
v2.0.0.0  2013-05-27 Added support for PID control of any data type by making PID class and functions templates. 
v1.1.0.0  2013-05-27 Added units tests and UnitTest++ library to test/. Made debug printing portable (by creating PrintDebug() function). Fixed bug with settingSampleMs for PidDbl object (added "this->"). Prevented derivative control for influencing the output until at least two calls to Run() have been made (i.e. doesn't assume previous input was 0.0 on first call to Run()).
v1.0.0.1  2013-05-27 Added README.rst.
v1.0.0.0  2012-10-01 Initial commit.
========= ========== ===================================================================================================

test/main.cpp
//!
//! @file 			main.cpp
//! @author 		Geoffrey Hunter <gbmhunter@gmail.com> (www.mbedded.ninja)
//! @created		2014-08-12
//! @last-modified 	2019-05-20
//! @brief 			Contains main entry point for unit tests.
//! @details
//!					See README.rst in root dir for more info.

//===== SYSTEM LIBRARIES =====//
// none

//====== USER LIBRARIES =====//
#include "MUnitTest/MUnitTestApi.hpp"

//===== USER SOURCE =====//
// none


int main()
{

	// Run all unit tests and return the result
	return MbeddedNinja::TestRegister::RunAllTests();
	
}


test/pidtest.cpp
//!
//! @file 			PidTest.cpp
//! @author 		Geoffrey Hunter <gbmhunter@gmail.com> (www.cladlab.com)
//! @edited 		n/a
//! @created		2013/05/27
//! @last-modified 	2014/03/24
//! @brief 			Unit test file for Pid library.
//! @details
//!					See README.rst in repo root dir for more info.

#include "../api/CP3id.hpp"

// UnitTest++ library
#include "UnitTest++/src/UnitTest++.h"

using namespace CP3id;


TEST(ZpZiZdCorrectTest)
{
	Pid<double> pidTest;

	double kp = 1.0;
	double ki = 5.0;
	double kd = 6.0;
	double timeStepMs = 1000.0;

	// Set-up controller with P, I and D values
	pidTest.Init(
		kp,									//!< Kp
		ki,									//!< Ki
		kd,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::VELOCITY_PID,		//!< Control type
		timeStepMs,									//!< Update rate (ms)
		0.0,										//!< Min output
		100.0,										//!< Max output
		0.0											//!< Initial set-point
	);
	
	// Now check that Kp, Ki, and Kd were calculated correctly
	
	CHECK_CLOSE(kp, pidTest.GetKp(), 0.0001);
	CHECK_CLOSE(ki, pidTest.GetKi(), 0.0001);
	CHECK_CLOSE(kd, pidTest.GetKd(), 0.0001);

	// Check Zp, Zi, and Zd (time-scaled Ki, Kd, Kp) were calculated correctly
	CHECK_CLOSE(kp, pidTest.GetZp(), 0.0001);
	CHECK_CLOSE(ki/(timeStepMs/1000.0), pidTest.GetZi(), 0.0001);
	CHECK_CLOSE(kd*(timeStepMs/1000.0), pidTest.GetZd(), 0.0001);
}

TEST(ZeroInZeroOutTest)
{
	Pid<double> pidTest;

	// Set-up P only controller
	pidTest.Init(
		1.0,									//!< Kp
		0.0,									//!< Ki
		0.0,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::VELOCITY_PID,		//!< Control type
		10.0,									//!< Update rate (ms)
		0.0,									//!< Min output
		100.0,								//!< Max output
		0.0									//!< Initial set-point
	);
	
	pidTest.Run(0.0);

	CHECK_CLOSE(0, pidTest.output, 0.0001);
}

TEST(POnlyPositiveTest)
{
	Pid<double> pidTest;

	// Set-up P only controller, non accumulating
	pidTest.Init(
		1.0,									//!< Kp
		0.0,									//!< Ki
		0.0,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::DONT_ACCUMULATE_OUTPUT,		//!< Control type
		1000.0,								//!< Update rate (ms)
		-100.0,									//!< Min output
		100.0,								//!< Max output
		0.0									//!< Initial set-point
	);
	
	// Perform one execution
	pidTest.Run(1.0);

	// Output should be -1
	CHECK_CLOSE(-1.0, pidTest.output, 0.0001);
}

TEST(POnlyNegativeTest)
{
	Pid<double> pidTest;

	// Set-up P only controller, non accumulating
	pidTest.Init(
		1.0,									//!< Kp
		0.0,									//!< Ki
		0.0,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::DONT_ACCUMULATE_OUTPUT,		//!< Control type
		1000.0,								//!< Update rate (ms)
		-100.0,									//!< Min output
		100.0,								//!< Max output
		0.0									//!< Initial set-point
	);
	
	// Perform one execution
	pidTest.Run(-1.0);

	// Output should be 1
	CHECK_CLOSE(1.0, pidTest.output, 0.0001);
}

TEST(IOnlyPositiveTest)
{
	Pid<double> pidTest;

	// Set-up P only controller, non accumulating
	pidTest.Init(
		0.0,									//!< Kp
		10.0,									//!< Ki
		0.0,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::DONT_ACCUMULATE_OUTPUT,		//!< Control type
		1000.0,								//!< Update rate (ms)
		-100.0,									//!< Min output
		100.0,								//!< Max output
		0.0									//!< Initial set-point
	);
	
	// Perform one execution
	pidTest.Run(1.0);

	// Output should be -10
	CHECK_CLOSE(-10.0, pidTest.output, 0.0001);

	// Perform one execution
	pidTest.Run(1.0);

	// Output should be -20
	CHECK_CLOSE(-20.0, pidTest.output, 0.0001);
}

TEST(IOnlyNegativeTest)
{
	Pid<double> pidTest;

	// Set-up P only controller, non accumulating
	pidTest.Init(
		0.0,									//!< Kp
		10.0,									//!< Ki
		0.0,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::DONT_ACCUMULATE_OUTPUT,		//!< Control type
		1000.0,								//!< Update rate (ms)
		-100.0,									//!< Min output
		100.0,								//!< Max output
		0.0									//!< Initial set-point
	);
	
	// Perform one execution
	pidTest.Run(-1.0);

	// Output should be 10
	CHECK_CLOSE(10.0, pidTest.output, 0.0001);

	// Perform one execution
	pidTest.Run(-1.0);

	// Output should be 20
	CHECK_CLOSE(20.0, pidTest.output, 0.0001);
}

TEST(DOnlyPositiveTest)
{
	Pid<double> pidTest;

	// Set-up P only controller, non accumulating
	pidTest.Init(
		0.0,									//!< Kp
		0.0,									//!< Ki
		1.0,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::DONT_ACCUMULATE_OUTPUT,		//!< Control type
		1000.0,								//!< Update rate (ms)
		-100.0,									//!< Min output
		100.0,								//!< Max output
		0.0									//!< Initial set-point
	);
	
	// Perform one execution
	pidTest.Run(1.0);

	// Output should be still 0 (at least two calls to Run()
	// required before derivative control starts working
	CHECK_CLOSE(0.0, pidTest.output, 0.0001);

	// Perform one execution
	pidTest.Run(1.0);

	// Output should be 0.0, since no change between the two
	// last calls to Run().
	CHECK_CLOSE(0.0, pidTest.output, 0.0001);

	// Perform one execution
	pidTest.Run(0.0);

	// Output should be 1.0
	CHECK_CLOSE(1.0, pidTest.output, 0.0001);
}

TEST(DOnlyNegativeTest)
{
	Pid<double> pidTest;

	// Set-up P only controller, non accumulating
	pidTest.Init(
		0.0,									//!< Kp
		0.0,									//!< Ki
		1.0,									//!< Kd
		Pid<double>::ControllerDirection::PID_DIRECT,		//!< Control type
		Pid<double>::OutputMode::DONT_ACCUMULATE_OUTPUT,		//!< Control type
		1000.0,								//!< Update rate (ms)
		-100.0,									//!< Min output
		100.0,								//!< Max output
		0.0									//!< Initial set-point
	);
	
	// Perform one execution
	pidTest.Run(-1.0);

	// Output should be still 0 (at least two calls to Run()
	// required before derivative control starts working
	CHECK_CLOSE(0.0, pidTest.output, 0.0001);

	// Perform one execution
	pidTest.Run(-1.0);

	// Output should be 0.0, since no change between the two
	// last calls to Run().
	CHECK_CLOSE(0.0, pidTest.output, 0.0001);

	// Perform one execution
	pidTest.Run(0.0);

	// Output should be -1.0
	CHECK_CLOSE(-1.0, pidTest.output, 0.0001);
}

int main()
{
	return UnitTest::RunAllTests();
}


src/Pid.cpp
//!
//! @file 			Pid.cpp
//! @author 		Geoffrey Hunter <gbmhunter@gmail.com> (wwww.mbedded.ninja)
//! @edited 		n/a
//! @created		2012-11-09
//! @last-modified 	2014-10-10
//! @brief 			PID library for embedded systems, written in C++.
//! @details
//!					See README.rst in repo root dir for more info.


//===============================================================================================//
//===================================== HEADER GUARD ============================================//
//===============================================================================================//

#ifndef __cplusplus
	#error Please build with C++ compiler
#endif

//===============================================================================================//
//========================================= INCLUDES ============================================//
//===============================================================================================//

//===== SYSTEM LIBRARIES =====//
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

//===== USER SOURCE =====//
#include "../include/Pid.hpp"

//===============================================================================================//
//================================== PRECOMPILER CHECKS =========================================//
//===============================================================================================//



//===============================================================================================//
//======================================== NAMESPACE ============================================//
//===============================================================================================//

namespace MbeddedNinja
{
	namespace MPidNs
	{

		// All PID code in Pid.hpp (templates)

	} // namespace MPidNs

} // namespace MbeddedNinja

// EOF


api/MPidApi.hpp
//!
//! @file 			MPidApi.hpp
//! @author 		Geoffrey Hunter <gbmhunter@gmail.com> (www.mbedded.ninja)
//! @edited 		n/a
//! @created		2014-03-24
//! @last-modified 	2014-03-24
//! @brief 			API header file for the MPid module.
//! @details
//!					See README.rst in repo root dir for more info.


//===============================================================================================//
//====================================== HEADER GUARD ===========================================//
//===============================================================================================//

#ifndef __cplusplus
	#error Please build with C++ compiler
#endif

#ifndef M_PID_M_PID_API_H
#define M_PID_M_PID_API_H

#include "../include/Pid.hpp"

#endif // #ifndef M_PID_M_PID_API_H

// EOF


include/Config.hpp
//!
//! @file 			Config.hpp
//! @author 		Geoffrey Hunter <gbmhunter@gmail.com> (www.cladlab.com)
//! @edited 		n/a
//! @created		2014/03/24
//! @last-modified 	2014/03/24
//! @brief 			Configuration file for the PID library.
//! @details
//!					See README.rst in repo root dir for more info.

//===============================================================================================//
//====================================== HEADER GUARD ===========================================//
//===============================================================================================//

#ifndef __cplusplus
	#error Please build with C++ compiler
#endif

#ifndef CP3ID_CONFIG_H
#define CP3ID_CONFIG_H

//===============================================================================================//
//================================== PRECOMPILER CHECKS =========================================//
//===============================================================================================//

//! @brief		Set to 1 if you want debug code enabled. Set to 0 if you want to save memory.
#define cp3id_config_INCLUDE_DEBUG_CODE			1

#if(cp3id_config_INCLUDE_DEBUG_CODE == 1)
	//! @brief		Size (in bytes) of the debug buffer (only valid if
	//!				cp3id_config_INCLUDE_DEBUG_CODE is set to 1).
	#define cp3id_config_DEBUG_BUFF_SIZE		200
#endif

#endif // #ifndef CP3ID_CONFIG_H

// EOF


include/Pid.hpp
//!
//! @file 			Pid.hpp
//! @author 		Geoffrey Hunter <gbmhunter@gmail.com> (www.mbedded.ninja)
//! @edited 		n/a
//! @created		2012-10-01
//! @last-modified 	2014-10-10
//! @brief
//! @details
//!					See README.rst in repo root dir for more info.

#ifndef __cplusplus
	#error Please build with C++ compiler
#endif

//===============================================================================================//
//====================================== HEADER GUARD ===========================================//
//===============================================================================================//


#ifndef M_PID_PID_H
#define M_PID_PID_H

//===== SYSTEM LIBRARIES =====//
#include <stdint.h>		// uint32_t
//#include <iostream>		//! @debug

//===== USER LIBRARIES =====//
//#include "MSlotmachine/api/MSlotmachineApi.hpp"			//!< Callback functionality

//===== USER SOURCE =====//
// none


namespace MbeddedNinja
{
	namespace MPidNs
	{
		
		//===============================================================================================//
		//===================================== CLASS DEFINITION ========================================//
		//===============================================================================================//

		//! @brief		PID class that uses dataTypes for it's arithmetic
		template <class dataType> class Pid
		{
			public:

				//===============================================================================================//
				//=================================== PUBLIC TYPEDEFS ===========================================//
				//===============================================================================================//

				//! @brief		Enumerates the controller direction modes
				enum class ControllerDirection
				{
					PID_DIRECT,			//!< Direct drive (+error gives +output)
					PID_REVERSE			//!< Reverse driver (+error gives -output)
				};

				//! Used to determine whether the output shouldn't be accumulated (distance control),
				//! or accumulated (velocity control).
				enum class OutputMode
				{
					DONT_ACCUMULATE_OUTPUT,
					ACCUMULATE_OUTPUT,
					DISTANCE_PID = DONT_ACCUMULATE_OUTPUT,
					VELOCITY_PID = ACCUMULATE_OUTPUT
				};
			
				//! @brief 		Init function
				//! @details   	The parameters specified here are those for for which we can't set up
				//!    			reliable defaults, so we need to have the user set them.
				Pid(
					dataType kp,
					dataType ki,
					dataType kd,
					ControllerDirection controllerDir,
					OutputMode outputMode,
					double samplePeriodMs,
					dataType minOutput,
					dataType maxOutput,
					dataType setPoint);

				//! @brief 		Computes new PID values
				//! @details 	Call once per sampleTimeMs. Output is stored in the pidData structure.
				void Run(dataType input);

				void SetOutputLimits(dataType min, dataType max);
			
				//! @details	The PID will either be connected to a direct acting process (+error leads to +output, aka inputs are positive)
				//!				or a reverse acting process (+error leads to -output, aka inputs are negative)
				void SetControllerDirection(ControllerDirection controllerDir);

				//! @brief		Changes the sample time
				void SetSamplePeriod(uint32_t newSamplePeriodMs);

				//! @brief		This function allows the controller's dynamic performance to be adjusted.
				//! @details	It's called automatically from the init function, but tunings can also
				//! 			be adjusted on the fly during normal operation
				void SetTunings(dataType kp, dataType ki, dataType kd);

				//! @brief		Returns the actual (not time-scaled) proportional constant.
				dataType GetKp();

				//! @brief		Returns the actual (not time-scaled) integral constant.
				dataType GetKi();

				//! @brief		Returns the actual (not time-scaled) derivative constant.
				dataType GetKd();

				//! @brief		Returns the time-scaled (dependent on sample period) proportional constant.
				dataType GetZp();

				//! @brief		Returns the time-scaled (dependent on sample period) integral constant.
				dataType GetZi();

				//! @brief		Returns the time-scaled (dependent on sample period) derivative constant.
				dataType GetZd();

				/*
				#if(cp3id_config_INCLUDE_DEBUG_CODE == 1)
					//! @brief		Pass in a callback for printing debug information.
					//! @details	Uses the slotmachine-cpp library to provide callback to method functionality.
					void SetDebugPrintCallback(SlotMachine::Callback<void, const char*> debugPrintCallback);
				#endif*/

				//! @brief 		The set-point the PID control is trying to make the output converge to.
				dataType setPoint;

				//! @brief		The control output.
				//! @details	This is updated when Pid_Run() is called.
				dataType output;
			
			private:

				/*
				#if(cp3id_config_INCLUDE_DEBUG_CODE == 1)
					//! @brief		Buffer for debug snprintf() calls.
					char debugBuff[cp3id_config_DEBUG_BUFF_SIZE];
				#endif*/

				//! @brief		Time-step scaled proportional constant for quick calculation (equal to actualKp)
				dataType Zp;

				//! @brief		Time-step scaled integral constant for quick calculation
				dataType Zi;

				//! @brief		Time-step scaled derivative constant for quick calculation
				dataType Zd;

				//! @brief		Actual (non-scaled) proportional constant
				dataType Kp;

				//! @brief		Actual (non-scaled) integral constant
				dataType Ki;

				//! @brief		Actual (non-scaled) derivative constant
				dataType Kd;

				//! @brief		Actual (non-scaled) proportional constant
				dataType prevInput;

				//! @brief		The change in input between the current and previous value
				dataType inputChange;
		
				//! @brief		The error between the set-point and actual output (set point - output, positive
				//! 			when actual output is lagging set-point.
				dataType error;

				//! @brief 		The output value calculated the previous time Pid_Run() was called.
				//! @details	Used in ACCUMULATE_OUTPUT mode.
				dataType prevOutput;

				//! @brief		The sample period (in milliseconds) between successive Pid_Run() calls.
				//! @details	The constants with the z prefix are scaled according to this value.
				double samplePeriodMs;

				dataType pTerm;				//!< The proportional term that is summed as part of the output (calculated in Pid_Run())
				dataType iTerm;				//!< The integral term that is summed as part of the output (calculated in Pid_Run())
				dataType dTerm;				//!< The derivative term that is summed as part of the output (calculated in Pid_Run())
				dataType outMin;				//!< The minimum output value. Anything lower will be limited to this floor.
				dataType outMax;				//!< The maximum output value. Anything higher will be limited to this ceiling.

				//! @brief		Counts the number of times that Run() has be called. Used to stop
				//!				derivative control from influencing the output on the first call.
				//! @details	Safely stops counting once it reaches 2^32-1 (rather than overflowing).
				uint32_t numTimesRan;

				//! @brief		The controller direction (FORWARD or REVERSE).
				ControllerDirection controllerDir;

				//! @brief		The output mode (non-accumulating vs. accumulating) for the control loop.
				OutputMode outputMode;
		};

		//===============================================================================================//
		//============================ TEMPLATE FUNCTION DEFINITIONS ====================================//
		//===============================================================================================//

		template <class dataType> Pid<dataType>::Pid(
			dataType kp,
			dataType ki,
			dataType kd,
			ControllerDirection controllerDir,
			OutputMode outputMode,
			double samplePeriodMs,
			dataType minOutput,
			dataType maxOutput,
			dataType setPoint) :
				numTimesRan(0)

		{
			//std::cout << __PRETTY_FUNCTION__ << " called." << std::endl;

			this->SetOutputLimits(minOutput, maxOutput);

			this->samplePeriodMs = samplePeriodMs;

			this->SetControllerDirection(controllerDir);
			this->outputMode = outputMode;
			
			// Set tunings with provided constants
			this->SetTunings(kp, ki, kd);
			this->setPoint = setPoint;
			this->prevInput = 0;
			this->prevOutput = 0;

			this->pTerm = 0.0;
			this->iTerm = 0.0;
			this->dTerm = 0.0;

		}

		template <class dataType> void Pid<dataType>::Run(dataType input)
		{
			// Compute all the working error variables
			//dataType input = *_input;
			
			//std::cout << __PRETTY_FUNCTION__ << " called with input = '" << input << "'." << std::endl;

			this->error = this->setPoint - input;
			
			// PROPORTIONAL CALCS

			this->pTerm = this->Zp*this->error;

			// INTEGRAL CALCS
			
			this->iTerm += (this->Zi * this->error);
			// Perform min/max bound checking on integral term
			if(this->iTerm > this->outMax)
				this->iTerm = this->outMax;
			else if(this->iTerm < this->outMin)
				this->iTerm = this->outMin;

			//===== DERIVATIVE CALS =====//

			// Only calculate derivative if run once or more already.
			if(this->numTimesRan > 0)
			{
				//std::cout << "numTimesRan is '" << this->numTimesRan << "'." << std::endl;
				this->inputChange = (input - this->prevInput);
				this->dTerm = -this->Zd*this->inputChange;
			}
			else
			{
				//std::cout << "numTimesRan is 0, derivative not calculated." << std::endl;
				this->dTerm = 0;
			}

			// Compute PID Output. Value depends on outputMode
			if(this->outputMode == OutputMode::DONT_ACCUMULATE_OUTPUT)
			{
				this->output =  this->pTerm + this->iTerm + this->dTerm;
			}
			else if(this->outputMode == OutputMode::ACCUMULATE_OUTPUT)
			{
				this->output = this->prevOutput + this->pTerm + this->iTerm + this->dTerm;
			}
			
			//std::cout << "this->output before limiting = '" << this->output << "'." << std::endl;

			// Limit output
			if(this->output > this->outMax)
				this->output = this->outMax;
			else if(this->output < this->outMin)
				this->output = this->outMin;
			
			// Remember input value to next call
			this->prevInput = input;
			// Remember last output for next call
			this->prevOutput = this->output;

			// Increment the Run() counter, after checking to make sure it hasn't reached
			// max value.
			if(this->numTimesRan < 0xFFFFFFFF)
				this->numTimesRan++;
		}

		//! @brief		Sets the PID tunings.
		//! @warning	Make sure samplePeriodMs is set before calling this funciton.
		template <class dataType> void Pid<dataType>::SetTunings(dataType kp, dataType ki, dataType kd)
		{
			if (kp<0 || ki<0 || kd<0)
				return;

			this->Kp = kp;
			this->Ki = ki;
			this->Kd = kd;

		   // Calculate time-step-scaled PID terms
		   this->Zp = kp;

			// The next bit requires double->dataType casting functionality.
		   this->Zi = ki * (dataType)(this->samplePeriodMs/1000.0);
		   this->Zd = kd / (dataType)(this->samplePeriodMs/1000.0);

		  if(this->controllerDir == ControllerDirection::PID_REVERSE)
		   {
			  this->Zp = (0 - this->Zp);
			  this->Zi = (0 - this->Zi);
			  this->Zd = (0 - this->Zd);
		   }
	/*
			#if(cp3id_config_INCLUDE_DEBUG_CODE == 1)
				snprintf(debugBuff,
					sizeof(debugBuff)/sizeof(debugBuff[0]),
					"PID: Tuning parameters set. Kp = %.1f, Ki = %.1f, Kd = %f.1, "
					"Zp = %.1f, Zi = %.1f, Zd = %.1f, with sample period = %.1fms\r\n",
					Kp,
					Ki,
					Kd,
					Zp,
					Zi,
					Zd,
					samplePeriodMs);
				this->PrintDebugInfo(this->debugBuff);
			#endif*/
		}

		template <class dataType> dataType Pid<dataType>::GetKp()
		{
			return this->Kp;
		}

		template <class dataType> dataType Pid<dataType>::GetKi()
		{
			return this->Ki;
		}

		template <class dataType> dataType Pid<dataType>::GetKd()
		{
			return this->Kd;
		}

		template <class dataType> dataType Pid<dataType>::GetZp()
		{
			return this->Zp;
		}

		template <class dataType> dataType Pid<dataType>::GetZi()
		{
			return this->Zi;
		}

		template <class dataType> dataType Pid<dataType>::GetZd()
		{
			return this->Zd;
		}
		
		template <class dataType> void Pid<dataType>::SetSamplePeriod(uint32_t newSamplePeriodMs)
		{
		   if (newSamplePeriodMs > 0)
		   {
			  dataType ratio  = (dataType)newSamplePeriodMs
							  / (double)this->samplePeriodMs;
			  this->Zi *= ratio;
			  this->Zd /= ratio;
			  this->samplePeriodMs = newSamplePeriodMs;
		   }
		}
	
		template <class dataType> void Pid<dataType>::SetOutputLimits(dataType min, dataType max)
		{
			if(min >= max)
				return;
			this->outMin = min;
			this->outMax = max;

		}

		template <class dataType> void Pid<dataType>::SetControllerDirection(ControllerDirection controllerDir)
		{
			if(controllerDir != this->controllerDir)
			{
				// Invert control constants
				this->Zp = (0 - Zp);
				this->Zi = (0 - Zi);
				this->Zd = (0 - Zd);
			}
		   this->controllerDir = controllerDir;
		}
/*
		#if(cp3id_config_INCLUDE_DEBUG_CODE == 1)
			template <class dataType> void Pid<dataType>::PrintDebugInfo(const char* msg)
			{
				// Execute the callback, passing in the message
				this->debugPrintCallback.Execute(msg);
			}

			template <class dataType>
			void Pid<dataType>::SetDebugPrintCallback(SlotMachine::Callback<void, const char*> debugPrintCallback)
			{
				this->debugPrintCallback = debugPrintCallback;
			}
		#endif*/

	} // namespace MPid
} // namespace MbeddedNinja

#endif // #ifndef M_PID_PID_H

// EOF
